src/content/en/updates/2019/02/hotpath-with-wasm.md

```diff
@@ -0,0 +1,490 @@
+## It's consistently fast, yo.
+
+{% include "web/_shared/contributors/surma.html" %}
+
+In my [previous](/web/updates/2018/03/emscripting-a-c-library)
+[articles](/web/updates/2019/01/emscripten-npm) I talked about how WebAssembly
+allows you to bring the library ecosystem of C/C++ to the web. One app that
+makes extensive use of C/C++ libraries is [squoosh](https://squoosh.app/), our
+web app that allows you compress images with a variety of codecs that have been
+compiled from C++ to WebAssembly.
+
+WebAssembly is a low-level virtual machine that runs the bytecode that is stored
+in `.wasm` files. This byte code is strongly typed and structured in such a way
+that it can be compiled and optimized for the host system much quicker than
+JavaScript can. WebAssembly provides an environment to run code that had
+sandboxing and embedding in mind from the very start.
+
+In my experience, most performance problems on the web are caused by forced
+layout and excessive paint but every now and then an app needs to do a
+computationally expensive task that takes a lot of time. WebAssembly can help
+here.
+
+## The Hot Path
+
+In squoosh we wrote a [JavaScript
+function](https://github.com/GoogleChromeLabs/squoosh/blob/edd2c51eb6d0676a2e7b7e974337d58cbf00f1d1/src/codecs/rotate/processor.ts)
+that rotates an image buffer by multiples of 90 degrees. While
+[OffscreenCanvas](/web/updates/2018/08/offscreen-canvas) would be ideal for
+this, it isn't supported across the browsers we were targeting, and a little
+[buggy in Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=906619).
+
+This function iterates over every pixel of an input image and copies it to a
+different position in the output image to achieve rotation. For a 4094px by
+4096px image (16 megapixels) it would need over 16 million iterations of the
+inner code block, which is what we call a "hot path". Despite that rather big
+number of iterations, Chrome, Firefox and Safari finish the task in 1 second or
+less. An acceptable duration for this type of interaction.
+
+    for (let d2 = d2Start; d2 >= 0 && d2 < d2Limit; d2 += d2Advance) {
+      for (let d1 = d1Start; d1 >= 0 && d1 < d1Limit; d1 += d1Advance) {
+        const in_idx = ((d1 * d1Multiplier) + (d2 * d2Multiplier));
+        outBuffer[i] = inBuffer[in_idx];
+        i += 1;
+      }
+    }
+
+In Edge, however, it takes over 8 seconds. There are
+no browser APIs involved that could affect performance, so this is just about
+raw execution speed. 8 seconds is much less acceptable. Enter WebAssembly.
+
+## WebAssembly for predictable performance
+
+In general, JavaScript and WebAssembly can achieve the same peak performance.
+However, in JavaScript it's often tricky to stay on the "fast path". One key
+benefit that WebAssembly offers is _predictable_ performance, even across
+browsers. The strict typing and low-level architecture allows for stronger
+assumptions and more optimizations during compilation. The function above is a
```

Comment:
"more optimizations during compilation" seems to contradict what is said earlier about JS and wasm having the same peak performance. 

How about replacing this sentence:

> The strict typing and low-level architecture allows for stronger
> assumptions and more optimizations during compilation.

...with something like:

> JavaScript compilers generate high-performance machine code based on _speculative assumptions_, but have to deoptimize when those assumptions are violated (aka. "falling off the fast path"). In WebAssembly, there is no deoptimization step, thanks to the stronger guarantees offered by its strict typing and low-level architecture. WebAssembly can be optimized without speculating at all.

cc @gahaas

Reply:
I don’t think it contradicts the initial statement: AFAIU, difference is that wasm allows more optimizations _during compilation_, while JavaScript will  observed the code _running_ for a bit to make assumptions about typings and _then_ it will apply stronger optimizations.

I’d like to avoid going to deep into compiler theory here in the article, so I am trying to keep this section simple. WDYT?

Reply:
I agree with @mathiasbynens that it's easy to misunderstand the text as it is, but also with @surma  that it would be good to keep this section simple. I cannot find a good formulation yet though.

Reply:
> In general, JavaScript and WebAssembly can achieve the same peak performance. However, for JavaScript this performance can only be achieved on the "fast path", and it's often tricky to stay on that "fast path". One key benefit that WebAssembly offers is predictable performance, even across browsers. This means that in WebAssembly there only exists one "fast path", and it cannot be left. The strict typing and low-level architecture allow stronger assumption so that WebAssembly code only has to be optimized once.

Maybe like this?

Reply:
I like that approach. I’ll work it into the doc with some light rephrasing. Thank you very much

Reply:
> I don’t think it contradicts the initial statement: AFAIU, difference is that wasm allows more optimizations _during compilation_, while JavaScript will observed the code _running_ for a bit to make assumptions about typings and _then_ it will apply stronger optimizations.

The "during compilation" is the confusing part, since the current text is conflating two different meanings of "compilation".

For JS, one meaning is "compiling bytecode" i.e. turning the AST into bytecode that can then start to run (in the interpreter). The second meaning is "compiling machine code" (i.e. what an optimizing compiler does for hot code).

For Wasm, you get a bytecode format from the start; the only meaning of "compiling" then is "compiling machine code" (i.e. the second meaning above).

It only makes sense to compare those two; anything else is apples vs. oranges. And there is not much to compare _during_ compilation, since JS and Wasm use the same optimizing compiler.

I agree we should avoid diving into compiler theory :) but I think we can clarify this section without digging too deep. I like @gahaas's phrasing (or something among those lines). 

Alternatively, perhaps some inspiration could be taken from [the Liftoff blog post](https://v8.dev/blog/liftoff#tiering-up-from-liftoff-to-turbofan), e.g.

> WebAssembly is statically typed, so the engine can generate optimized code right away.

Reply:
Looks like this wasn't resolved yet. Also, the same statement appears in the intro of the article --- let's fix it there as well?